Adding tidal potential to gravity model

hyunseong96 opened the pull request:
Apologies for opening this PR so late. Antoniette and I took some time to check sign convention for the equation.
I open this PR to add gravity model plugin where I can add tidal potential exerted as satellite orbits around the host planet.

The equation of tidal potential is from Tobie et al. (2025) (https://doi.org/10.1007/s11214-025-01136-y).
<img width="612" alt="image" src="https://github.com/user-attachments/assets/97ae041e-00d9-491b-8981-6b4e254ec8f9" />
I only use T* and T0 because ep 4, 5 and 6 are dependent to short time scale (~3.55 days) and they are replaced by tidal heating.
T* is flattening and T0 is nonsynchronous rotation. T0 happens because ice surface is decoupled from rocky interior by the subsurface ocean and have difference rotational velocity.

To implement in cartesian coordinates, these equations are converted to ones using x, y and z instead of spherical variables.
which ends up as: 
dV_T/dx = (3* G *M_p)/(2 a^3) (1/6*(2*x)+1/2*(2*C1*x - 2*C2*y))
dV_T/dy = (3* G *M_p)/(2 a^3) (1/6*(2*y)+1/2*(2*C1*y - 2*C2*x))
dV_T/dz = (3* G *M_p)/(2 a^3) (1/6*(-4*z)).
where G is gravitational constant, M_p is mass of perturbing body, a is semi-major axis, C1 = cos(2*b*t) and C2 = sin(2*b*t).

This $V_T$ is from geodesy research, which takes potential and acceleration always positive as unit vector is inward to center of body. Therefore, for us, where unit vector is outward from the center of body, the return of plugin is $-g\vec{r}-\nabla(-V_T)\vec{r}$

test model is just to see how gravity changes with time. temperature does not change with time. viscosity is high throughout the model so there is no convection. Colorbar is for gravity magnitude, and arrow denotes direction of tidal acceleration.
<img width="613" height="449" alt="image" src="https://github.com/user-attachments/assets/56d5860f-be39-489d-bd58-78546776b30e" />

I will edit some parts quickly in test .prm, make more clarification in .cc and add change file!

hyunseong96 committed:
initiate gravity model for tidal effect

hyunseong96 commented:
Corrections are done but I am not sure why I have failure from dealii 9.5 test.
Is it okay if I just use git apply diff file?

bangerth commented:
@hyunseong96 Thank you for this patch! I will look at it later this week, but in the meantime, can you take a look at the test failure? It currently reports this:
```
******* Check /__w/aspect/aspect/build/tests/output-gravity_tidal_potential/statistics /__w/aspect/aspect/tests/gravity_tidal_potential/statistics
******* First 50 of 93 lines of diffs are:
----------------
##23      <==  0 0.000000000000e+00 0.000000000000e+00 96 3060 970 0 110 112 112 2.04033429e-05 7.15757443e-05 1.00000000e+02 9.99999535e+01 1.00000000e+02 output-gravity_tidal_potential/solution/solution-00000 -5.22786493e+05 5.82904414e+07 1.20123467e+08 9.17000000e+02 1.00000000e+20 2.63118556e+21 
##23      ==>  0 0.000000000000e+00 0.000000000000e+00 96 3060 970 0 95 96 96 2.04033436e-05 7.15759315e-05 1.00000000e+02 9.99999535e+01 1.00000000e+02 output-gravity_tidal_potential/solution/solution-00000 -5.22786377e+05 5.82904414e+07 1.20123467e+08 9.17000000e+02 1.00000000e+20 2.63118556e+21 

##23      #:8   <== 110
##23      #:8   ==> 95
@ Absolute error = 1.5000000000e+1, Relative error = 1.5789473684e-1
##23      #:9   <== 112
##23      #:9   ==> 96
@ Absolute error = 1.6000000000e+1, Relative error = 1.6666666667e-1
##23      #:10  <== 112
##23      #:10  ==> 96
@ Absolute error = 1.6000000000e+1, Relative error = 1.6666666667e-1
##23      #:12  <== 7.15757443e-05
##23      #:12  ==> 7.15759315e-05
@ Absolute error = 1.8720000000e-10, Relative error = 2.6154111540e-6
##23      #:17  <== -5.22786493e+05
##23      #:17  ==> -5.22786377e+05
@ Absolute error = 1.1600000000e-1, Relative error = 2.2188795482e-7
----------------
##24      <==  1 1.000000000000e+03 1.000000000000e+03 96 3060 970 0  50  51  51 2.53870595e-05 8.60011863e-05 1.00000000e+02 9.99999535e+01 1.00000000e+02 output-gravity_tidal_potential/solution/solution-00001 -5.23573285e+05 5.82904414e+07 1.20105088e+08 9.17000000e+02 1.00000000e+20 2.63118556e+21 
##24      ==>  1 1.000000000000e+03 1.000000000000e+03 96 3060 970 0 50 51 51 2.53870602e-05 8.60009790e-05 1.00000000e+02 9.99999535e+01 1.00000000e+02 output-gravity_tidal_potential/solution/solution-00001 -5.23573278e+05 5.82904414e+07 1.20105088e+08 9.17000000e+02 1.00000000e+20 2.63118556e+21 

##24      #:12  <== 8.60011863e-05
##24      #:12  ==> 8.60009790e-05
@ Absolute error = 2.0730000000e-10, Relative error = 2.4104376765e-6
##24      #:17  <== -5.23573285e+05
##24      #:17  ==> -5.23573278e+05
@ Absolute error = 7.0000000000e-3, Relative error = 1.3369666280e-8
----------------
##25      <==  2 2.000000000000e+03 1.000000000000e+03 96 3060 970 0  50  51  51 2.21503532e-05 8.26962512e-05 1.00000000e+02 9.99999535e+01 1.00000000e+02 output-gravity_tidal_potential/solution/solution-00002 -5.22786489e+05 5.82904414e+07 1.20011855e+08 9.17000000e+02 1.00000000e+20 2.63118556e+21 
##25      ==>  2 2.000000000000e+03 1.000000000000e+03 96 3060 970 0 50 51 51 2.21503531e-05 8.26962710e-05 1.00000000e+02 9.99999535e+01 1.00000000e+02 output-gravity_tidal_potential/solution/solution-00002 -5.22786367e+05 5.82904414e+07 1.20011855e+08 9.17000000e+02 1.00000000e+20 2.63118556e+21 

##25      #:17  <== -5.22786489e+05
##25      #:17  ==> -5.22786367e+05
@ Absolute error = 1.2200000000e-1, Relative error = 2.3336492246e-7
----------------
##27      <==  4 4.000000000000e+03 1.000000000000e+03 96 3060 970 0  50  51  51 2.49936170e-05 8.22470390e-05 1.00000000e+02 9.99999535e+01 1.00000000e+02 output-gravity_tidal_potential/solution/solution-00004 -5.22786489e+05 5.82904414e+07 1.20075508e+08 9.17000000e+02 1.00000000e+20 2.63118556e+21 
##27      ==>  4 4.000000000000e+03 1.000000000000e+03 96 3060 970 0 50 51 51 2.49936171e-05 8.22468386e-05 1.00000000e+02 9.99999535e+01 1.00000000e+02 output-gravity_tidal_potential/solution/solution-00004 -5.22786367e+05 5.82904414e+07 1.20075507e+08 9.17000000e+02 1.00000000e+20 2.63118556e+21 

##27      #:12  <== 8.22470390e-05
##27      #:12  ==> 8.22468386e-05
@ Absolute error = 2.0040000000e-10, Relative error = 2.4365678172e-6
##27      #:17  <== -5.22786489e+05
##27      #:17  ==> -5.22786367e+05
@ Absolute error = 1.2200000000e-1, Relative error = 2.3336492246e-7
----------------
##28      <==  5 5.000000000000e+03 1.000000000000e+03 96 3060 970 0  50  51  51 2.0495[945](https://github.com/geodynamics/aspect/actions/runs/16318503026/job/46090091115?pr=6614#step:6:946)9e-05 7.43142701e-05 1.00000000e+02 9.99999535e+01 1.00000000e+02 output-gravity_tidal_potential/solution/solution-00005 -5.22786489e+05 5.82904414e+07 1.20118224e+08 9.17000000e+02 1.00000000e+20 2.63118556e+21 
##28      ==>  5 5.000000000000e+03 1.000000000000e+03 96 3060 970 0 50 51 51 2.0495[946](https://github.com/geodynamics/aspect/actions/runs/16318503026/job/46090091115?pr=6614#step:6:947)0e-05 7.43144417e-05 1.00000000e+02 9.99999535e+01 1.00000000e+02 output-gravity_tidal_potential/solution/solution-00005 -5.22786369e+05 5.82904414e+07 1.20118224e+08 9.17000000e+02 1.00000000e+20 2.63118556e+21 

ninja: build stopped: subcommand failed.

CMake Error at /__w/aspect/aspect/tests/run_test.cmake:14 (message):
  *** Test aborted.
```

gassmoeller commented on include/aspect/gravity_model/radial.h:
```diff
@@ -135,6 +135,52 @@ namespace aspect
+
+
+    template <int dim>
+    class RadialLinearWithTidalPotential : public RadialLinear<dim>
```
While it works to derive a plugin from another plugin if you want to use the behavior of the parent class in the child class, I have noted over the years that it is rarely a good idea in ASPECT. It creates complicated class structures, and sometimes it straight out prevents making full use of the parent class (in particular if the parent has `protected` member variables or is itself derived from another class like here). I would recommend the following approach instead:

```suggestion
    class RadialLinearWithTidalPotential : public Interface<dim>, public SimulatorAccess<dim>
```

Then give this class a private member variable of type `RadialLinear<dim>` as a "base_gravity_model" and use its functionality by calling functions of this member variable. You can see how this works in the class `initial_temperature/patch_on_S40RTS.h` and its member variable `S40RTSPerturbation<dim> s40rts;`. Note in particular the lines 166-168 in `initial_temperature/patch_on_S40RTS.cc` where your class would initialize the member variable and let it read its own input parameters and line 76 where it uses its function `get_Vs`.

For now this is just a way to achieve the same functionality as your current approach (just with a bit more code, but easier to understand concept). But in the future this allows to expand the plugin to work with more than just the radial linear gravity model. For example the material model `prescribed_viscosity` can use an arbitrary material model as base model, and then modify it. In the same way your plugin could use any of the radial gravity profiles and apply a tidal potential on top.

gassmoeller commented on include/aspect/gravity_model/radial.h:
```diff
@@ -135,6 +135,52 @@ namespace aspect
+
+
+
+    template <int dim>
```
Please add doxygen documentation above the class name about what this class does. You can use the text you wrote at the bottom of the source file, but bring it into the doxygen format (/** ... */). This will be used in the online documentation (here the example for `RadialLinear`: https://aspect.geodynamics.org/doc/doxygen/classaspect_1_1GravityModel_1_1RadialLinear.html.

gassmoeller commented on include/aspect/gravity_model/radial.h:
```diff
@@ -135,6 +135,52 @@ namespace aspect
+        double b_NSR;
+
+        /**
+         * Radius of modelling satellite
```
By satellite you mean the moon/planet this gravity is computed for? This would be worth spelling out, since it may not be obvious for geoscientists focused on Earth.

gassmoeller commented on include/aspect/gravity_model/radial.h:
```diff
@@ -135,6 +135,52 @@ namespace aspect
+        double a_s;
+
+        /**
+         * Rate of non-synchronous rotation in m/year
```
Please use SI units wherever possible. You can still allow m/yr as input parameter if the global parameter `Use years instead of seconds` is set, and then convert into m/s when reading and storing the parameter.

And why m as length unit? Wouldnt it be much easier to ask the user to compute the angular velocity themselves beforehand and use that as input? Then you dont need to input the radius of the satellite either, or do you need that for anything else?

gassmoeller commented on include/aspect/gravity_model/radial.h:
```diff
@@ -135,6 +135,52 @@ namespace aspect
+
+      private:
+        /**
+         * Mass of hosting planet
```
what is a "hosting" planet? Do you mean the mass of the planet/star that generates the tidal potential? That would be worth spelling out.

gassmoeller commented on include/aspect/gravity_model/radial.h:
```diff
@@ -135,6 +135,52 @@ namespace aspect
+        double M_p;
+
+        /**
+         * Semimajor axis of satellite's orbit
```
I ask this below already, but by "satellite" you mean the object ASPECT is modelling? Please spell that out.

gassmoeller commented on source/gravity_model/radial.cc:
```diff
@@ -178,6 +183,159 @@ namespace aspect
+    Tensor<1,dim>
+    RadialLinearWithTidalPotential<dim>::gravity_vector (const Point<dim> &p) const
+    {
+      /**
+       * This gravity model is a plugin for a case where the
+       * tidal potential by flattening and non-synchronnous rotation changes gravity with position and time.
+       *
+       * The equation implemented in this heating model is from Tobie et al. (2025) (https://doi.org/10.1007/s11214-025-01136-y),
+       * which is defined as:
+       * g = -magnitude - gradient (-tidal potential)).
+       * potential = (3 G M_p) / (2 a_s^3) * r^2 * (Tstar + T0)
+       * Tstar = 1/6 *(1-3*cos(theta)^2) and T0=1/2sin(theta)^2*cos(2*lambda + 2*b*t)
+       * where G = gravitational constant, M_p = mass of planet, a_s = semimajor axis of satellite's orbit, b = angular rate of nonsynchronous rotation.
+       * r, theta and lambda are radial distance, polar angle and azimuthal angle, respectively.
+       * b [1/s] = b_NSR [m/yr] / (circumference of satellite [m]) / year_to_seconds [s/yr] * 2 * pi
```
This part of the explanation needs to go to the header file where I ask for documentation about this class. It is a high level summary of what this plugin is for.

The part below my comment (about conversion to Cartesian coordinates) can stay here, it is only relevent for the implementation, but not for understanding what the purpose of this plugin is.

gassmoeller commented on source/gravity_model/radial.cc:
```diff
@@ -178,6 +183,159 @@ namespace aspect
+    {
+      AssertThrow (dim==3, ExcMessage ("The 'radial linear with tidal potential' gravity model "
+                                       "can only be used in 3D."));
+      this->RadialLinear<dim>::parse_parameters (prm);
```
this line will change with my comments about using a member object of type `RadialLinear`.

gassmoeller commented on source/gravity_model/radial.cc:
```diff
@@ -178,6 +183,159 @@ namespace aspect
+
+      prm.enter_subsection("Gravity model");
+      {
+        prm.enter_subsection("Radial linear with tidal potnetial");
```
typo in the section name. Also just use "Radial with tidal potential". That allows extending the plugin later.

gassmoeller commented on source/gravity_model/radial.cc:
```diff
@@ -178,6 +183,159 @@ namespace aspect
+        {
+          M_p = prm.get_double ("Mass of planet");
+          a_s = prm.get_double ("Semimajor axis of satellite");
+          b_NSR = prm.get_double ("Rate of nonsynchronous rotation");
```
here you can do the following to allow for different time units in the input file, but still use SI units in the code:

```suggestion
          const double time_scale = this->get_parameters().convert_to_years ? constants::year_in_seconds : 1.0;
          b_NSR = prm.get_double ("Rate of nonsynchronous rotation") / time_scale;
```

gassmoeller commented on source/gravity_model/radial.cc:
```diff
@@ -178,6 +183,159 @@ namespace aspect
+        {
+          prm.declare_entry ("Mass of planet", "1.898e27",
+                             Patterns::Double (),
+                             "Mass of satellte's host planet. "
```
```suggestion
                             "Mass of satellite's host planet. "
```

gassmoeller commented on source/gravity_model/radial.cc:
```diff
@@ -178,6 +183,159 @@ namespace aspect
+                             "Units is $kg$.");
+          prm.declare_entry ("Semimajor axis of satellite", "670900000",
+                             Patterns::Double (),
+                             "Length of semimajor axis of satellite. "
```
Better to be more verbose:
 ```suggestion
                             "Length of the semimajor axis of the orbit of the modeled planet. "
```

gassmoeller commented on source/gravity_model/radial.cc:
```diff
@@ -178,6 +183,159 @@ namespace aspect
+    void
+    RadialLinearWithTidalPotential<dim>::parse_parameters (ParameterHandler &prm)
+    {
+      AssertThrow (dim==3, ExcMessage ("The 'radial linear with tidal potential' gravity model "
```
update name

gassmoeller commented on source/gravity_model/radial.cc:
```diff
@@ -178,6 +183,159 @@ namespace aspect
+        prm.enter_subsection("Radial linear with tidal potnetial");
+        {
+          M_p = prm.get_double ("Mass of planet");
+          a_s = prm.get_double ("Semimajor axis of satellite");
```
The current spelling is easy to confuse, if you model an elliptical body on an elliptical orbit.
```suggestion
          a_s = prm.get_double ("Semimajor axis of orbit");
```

gassmoeller reviewed (commented):
I will go ahead and leave some comments, Wolfgang can of course chime in on the mathematics (I havent compared the implemented formula with the paper). Some comments on the code and structure below. I think this is a valuable addition, but I would change the structure of the code a bit. Let me know in case you have questions or want some guidance on how to do something.

hyunseong96 commented:
@gassmoeller Hi Rene, thanks for reviewing the PR. I have few questions.
First, when you mentioned using angular rate instead of m/year. I thought about just using angular rate as 1/s, but seeing the suggestion below confused me a bit.
> const double time_scale = this->get_parameters().convert_to_years ? constants::year_in_seconds : 1.0;
          b_NSR = prm.get_double ("Rate of nonsynchronous rotation") / time_scale;

Did you mean the parameter, b_NSR, as angle/year? or is there any perspective I am not aware of inside the code while solving equations?



Second, separating into different files seems a bit risky with my poor skill of git commands sadly. I found out that I need to add this line to radial constant plugin.
`template <int dim2> friend class RadialWithTidalPotential;`

After the incident of having hundred of others' commits on my PR at Hackathon, I am still not sure how I can update the branch so I can adjust my plugin to the update you will do about radial plugins. I searched for this and tried on branch I am not using anymore.

> Sync fork of my repository (main) at github - git remote update - git checkout <branch> - git merge origin/main

Would this four steps be valid to keep up with updates? Otherwise I think it can be easier for me to close the PR and open a new one after your radial separation PR is merged.

gassmoeller commented:
>  First, when you mentioned using angular rate instead of m/year. I thought about just using angular rate as 1/s, but seeing the suggestion below confused me a bit.
> Did you mean the parameter, b_NSR, as angle/year? or is there any perspective I am not aware of inside the code while solving equations?

Yes I could have been more clear in my comment there. In general there are two separate issues:
1. I think expressing the rotation velocity in angular velocity instead of velocity makes more sense, because then you do not need the input of the satellite radius and the internal conversion to angular velocity.
2. In many places in ASPECT we expect that the input of velocities is given as unit per second (if the parameter `Use years instead of seconds` is `false`), or as unit per years (if the parameter is `true`). But internally we always want to use consistent SI units, so if the velocity is given by the input file as `1/yr` then after reading it we need to convert to `1/s`. This is all I wanted to express with my suggestions. The code I mentioned should do exactly that: Read the input, then if the input is given in years instead of seconds scale it with a time scale and convert to the proper SI unit 1/s:
```
const double time_scale = this->get_parameters().convert_to_years ? constants::year_in_seconds : 1.0;
b_NSR = prm.get_double ("Rate of nonsynchronous rotation") / time_scale;
```

> I found out that I need to add this line to radial constant plugin.
`template <int dim2> friend class RadialWithTidalPotential;`

I do not see why you would need to do this. You can declare a class (in this `RadialConstant`) as a `friend` of another class (in this case `RadialWithTidalPotential`), which allows the other class to access private member variables of the first class. But I do not see where you do this, the only operation your new class needs of `RadialConstant` is to call its public member functions `declare_parameters`, `parse_parameters` and `gravity_vector`. No access to private variables is needed. Can you upload your full new code following the instructions I give below?

> After the incident of having hundred of others' commits on my PR at Hackathon, I am still not sure how I can update the branch so I can adjust my plugin to the update you will do about radial plugins. I searched for this and tried on branch I am not using anymore.

>    Sync fork of my repository (main) at github - git remote update - git checkout - git merge origin/main

> Would this four steps be valid to keep up with updates? Otherwise I think it can be easier for me to close the PR and open a new one after your radial separation PR is merged.

Please do not use this workflow, it is what caused all the commits to appear in your branch. It is a valid git workflow and some projects use it, but we do not. I could explain our alternate workflow and the reasons for it (using an operation called `git rebase`), but it is not necessary for what you want to do here. No update from the main branch is necessary, just continue on your branch and move all the new code you added into a new file and remove it from radial.h/.cc. Afterwards, there should not be any changes in radial.h/.cc left and therefore it will not conflict with my changes on the main branch. We will have to squash all your commits before merging this PR in the end, but I will write you instructions for it when we are at that stage. For now just keep making changes on this branch without updates from main.

gassmoeller reviewed (commented):
Yes, much better structure. I left a number of comments, but they are mostly suggestions for improvements, I didnt see anything wrong.

Based on the error messages that you posted and your questions, I think it is worth for you to review the difference between static functions and non static functions of a class in c++. Here is a short summary: https://www.learncpp.com/cpp-tutorial/static-member-functions/.

In short, 'normal' member functions of a class (say class `RadialConstant`) require an object of this class to be created (in your case `radialconstant`). Then you can call the member functions like so: `radialconstant.gravity_vector(...)`. Static member functions however do not require an object of that class, and are usually called through their class name (instead of a particular object name), like so: `RadialConstant<dim>::declare_parameters(...)`.

hyunseong96 commented on source/gravity_model/radial.cc:
```diff
@@ -25,6 +25,11 @@
 #include <deal.II/base/tensor.h>
 
+#include <aspect/geometry_model/spherical_shell.h>
+#include <aspect/geometry_model/sphere.h>
+#include <aspect/geometry_model/chunk.h>
+#include <aspect/geometry_model/ellipsoidal_chunk.h>
+
```
I think you are right. My initial thought was my plugin should make error when spherical geometry is not used, but Radial constant would already do this. I will change it!

gassmoeller commented on include/aspect/gravity_model/radial_with_tidal_potential.h:
```diff
@@ -0,0 +1,96 @@
+        double M_p;
+
+        /**
+         * Semimajor axis of the modeled body
```
What do you think about: `Semimajor axis of the orbit that causes the tidal perturbation`? More explanation in the parameter description is of course also welcome.

hyunseong96 commented:
@gassmoeller Hi Rene, I finished corrections by your suggestions and plugin works as planned!
Just one thing is that, after discussion with @Geoniette , we decided that it looks better to use period in year/second than angular rate as input. Then, period is converted to angular frequency(rate) in code.
This change was made in [dc77c71](https://github.com/geodynamics/aspect/pull/6614/commits/dc77c7161d38b8da9908256d0ba9d18e0eeffbdd).
If code looks all fine, then I can make rebase to reduce number of commits!
Thank you :)

gassmoeller reviewed (commented):
Very nice. Just a few minor comments left. Yes, please rebase and squash the commits.

hyunseong96 commented:
Hi @gassmoeller ,

sorry for taking this so long.
Here is the issue I am facing after changing into accessing private member, radialconstant.
In radial constant gravity plugin, there is an AssertThrow for checking geometry.
`AssertThrow (this->get_geometry_model().natural_coordinate_system() == Utilities::Coordinates::spherical ||
                   this->get_geometry_model().natural_coordinate_system() == Utilities::Coordinates::ellipsoidal,
                   ExcMessage ("Gravity model 'radial constant' should not be used with geometry models that "
                               "do not have either a spherical or ellipsoidal natural coordinate system."));`
However, with my code, it is not possible to read the geometry.

when I removed AssertThrow to try whether my code works anyways, I saw that magnitude was chosen from default value, although I put magnitude value in .prm file.

If not a problem, could I remove the lines for private member and copy&paste the lines needed from radial constant?
The initial thought, when Wolfgang, Antoniette and I had discussion, was that it is easier to inherit from existing plugins, but now it is facing some issues.

hyunseong96 commented:
Hi @gassmoeller ,
Now gravity related to nonsynchronous rotation works well as intended with radial constant. 
Here I squashed commits too!
Cheers, Hyunseong

gassmoeller commented on include/aspect/gravity_model/radial_with_tidal_potential.h:
```diff
@@ -0,0 +1,95 @@
+
+      private:
+        /**
+         *
```
This variable needs documentation, e.g.


```suggestion
         * The radially constant part of the gravity is handled by an object of type RadialConstant.
```

gassmoeller commented on source/gravity_model/radial_with_tidal_potential.cc:
```diff
@@ -0,0 +1,155 @@
+
+      const double angular_frequency = 2. * dealii::numbers::PI / P;
+      const double C1 = std::cos( 2. * angular_frequency * t);
+      const double C2 = std::sin( 2. * angular_frequency * t);
```
If the gravity model depends on the current model time `t` does that mean your ASPECT model has to resolve the rotational period of the moon around its planet? I.e. does this model only give reasonable gravity values if the the ASPECT timestep value is at most a few days (or weeks)? Does this need to be taken into account when choosing ASPECTs time step? Then document this, or assert that the ASPECT time step is not larger than the orbital period (or probably significantly shorter than that).

hyunseong96 replied:
Yes and I put the name ambiguously. It is an angular period of the nonsynchronous rotation (NSR). Europa's NSR period is in an order of ~10,000 years, but you are right that model calculation time steps should be smaller than the period of NSR.

so I changed the name from angular_frequency to NSR_angular_frequency and added assertthrow for maximum time step being larger than the period of NSR

gassmoeller commented on tests/gravity_tidal_potential.prm:
```diff
@@ -0,0 +1,99 @@
+
+subsection Gravity model
+  set Model name = radial with tidal potential
+  subsection Radial constant
+    set Magnitude = 1.3
+  end
+  subsection Radial with tidal potential
+  end
```
Please add some parameters here, even if they just set all parameters to the default values. This will show users how to use this gravity model.

gassmoeller reviewed (commented):
Yes this looks almost ready now. I have some minor comments and one question I would like to understand better. Let me know when this is ready for another look.

bangerth commented:
I missed this PR. I will gladly also review it -- tag me when you've addressed @gassmoeller 's suggestions!

hyunseong96 commented:
Hi both! @gassmoeller @bangerth 

I have corrected following suggestions. 
I added assertthrow to return error when maximum time step is larger than the period of nonsynchronous rotation.
Thank you for your help and reviewing! :)

bangerth commented on include/aspect/gravity_model/radial_with_tidal_potential.h:
```diff
@@ -0,0 +1,95 @@
+     * Tstar = 1/6 *(1-3*cos(theta)^2) and T0=1/2sin(theta)^2*cos(2*lambda + 2*b*t)
+     * where G = gravitational constant, M_p = mass of the perturbing body, a_s = semimajor axis of the orbit, b = angular rate of non-synchronous rotation.
+     * b = 2 * pi / P where P is period of NSR.
+     * r, theta and lambda are radial distance, polar angle and azimuthal angle, respectively.
```
This is going to be difficult to read in a browser where line breaks might be at different positions. Let's do it like this:

```suggestion
     * which is defined as:
     * $\mathbf g = -\text{magnitude} - \nabla (-\phi)$.
     * Tidal potential $\phi$ is positive because the formula follows conventions from geodesy research, where potential is taken as positive. We use
     * $\phi = \frac{3 G M_p}{2 a_s^3} r^2 (T^\ast + T_0)$ where
     * $T^\ast = \frac{1}{6} (1-3 \cos(\theta)^2)$ and $T_0=\frac{1}{2}\sin(\theta)^2 \cos(2\lambda + 2bt)$
     * where $G$ is the gravitational constant, $M_p$ the mass of the perturbing body, $a_s$ the semimajor axis of the orbit, $b$ the angular rate of non-synchronous rotation, and
     * $b = 2  \frac{\pi}{P} where $P$ is period of NSR.
     * $r$, $\theta$ and $\lambda$ are radial distance, polar angle, and azimuthal angle, respectively.
```
Separately, can you spell out what NSR is? It's an abbreviation not previously explained here.

hyunseong96 replied:
I somehow can't make a comment on the review about addressing Tstar and T0. 
However, I leave here because I think comment here is also related.
Tstar is static flattening of the satellite,.
T0 is the quasi-static stretching in the direction of the planet, which slowly vary in nonsynchronous rotation.
These two are the components for nonsynchronous rotation (Tobie et al., 2025; Wahr et al. 2009)!

bangerth commented on include/aspect/gravity_model/radial_with_tidal_potential.h:
```diff
@@ -0,0 +1,95 @@
+         * Read the parameters this class declares from the parameter file.
+         */
+        void
+        parse_parameters (ParameterHandler &prm) override;
+
+      private:
+        /**
+         * The radially constant part of the gravity is handled by an object of type RadialConstant.
+         */
+        RadialConstant<dim> radialconstant;
+
+        /**
+         * Mass of the perturbing body
+         */
+        double M_p;
+
+        /**
+         * Semimajor axis of the orbit that causes the tidal perturbation
+         */
+        double a_s;
+
+        /**
+         * Period of the non-synchronous rotation
+         */
+        double P;
```
Can you annotate these variables in the style of https://github.com/geodynamics/aspect/pull/7112/changes?

bangerth replied:
I specifically meant the last three members.

bangerth commented on source/gravity_model/radial_with_tidal_potential.cc:
```diff
@@ -0,0 +1,150 @@
+
+      const Tensor<1,dim> dTstar_gradient ({1./3. * p[0], 1./3. * p[1], -2./3. * p[2]});
+
+      const Tensor<1,dim> dT0_gradient ({C1 *p[0] - C2 *p[1], -C1 *p[1] - C2 *p[0], 0});
```
Can you still address this?

bangerth commented on source/gravity_model/radial_with_tidal_potential.cc:
```diff
@@ -0,0 +1,158 @@
+                             Patterns::Double (),
+                             "Mass of body that perturbs gravity of modeled body. "
+                             "The default value is chosen for modeling Europa, therefore, it is the mass of Jupiter. "
+                             "Units is $kg$.");
```
```suggestion
                             "Units: \si{\kilo\gram}.");
```

bangerth commented on source/gravity_model/radial_with_tidal_potential.cc:
```diff
@@ -0,0 +1,158 @@
+                             "For example, tidal perturbation on Europa happens by Europa orbiting Jupiter, "
+                             "and that on Earth, if Moon is in consideration, happens by Moon orbiting Earth. "
+                             "The default value is for the semimajor axis of Europa's orbit. "
+                             "Units is $m$.");
```
```suggestion
                             "Units: \si{meter}.");
```

bangerth commented on source/gravity_model/radial_with_tidal_potential.cc:
```diff
@@ -0,0 +1,158 @@
+                             "Period of nonsynchronous rotation (NSR). "
+                             "The default value is the period of NSR on Europa's icy shell. "
+                             "Units is $year$ when 'Use years instead of seconds' is true, "
+                             "and $second$ when 'Use years instead of seconds' is false. ");
```
```suggestion
                             "Units is \si{year} when 'Use years instead of seconds' is true, "
                             "and \si{second} when 'Use years instead of seconds' is false. ");
```

hyunseong96 replied:
I had a quick look at the \si{} does not have year as time unit according to siunitx. so I did not use \si{} for year.

bangerth replied:
Ah, ok, thanks for checking!

bangerth commented on source/gravity_model/radial_with_tidal_potential.cc:
```diff
@@ -0,0 +1,158 @@
+          P = prm.get_double ("Period of nonsynchronous rotation") * time_scale;
+
+          AssertThrow (this->get_parameters().maximum_time_step < P , ExcMessage ("Time step size should be smaller than the period of nonsynchronous rotation. "
+                                                                                  "Otherwise, the tidal potential will not be resolved properly."));
```
Let's make these lines shorter:

```suggestion
          AssertThrow (this->get_parameters().maximum_time_step < P , 
                                  ExcMessage ("Time step size should be smaller than the period of nonsynchronous rotation. "
                                                          "Otherwise, the tidal potential will not be resolved properly."));
```

bangerth commented on source/gravity_model/radial_with_tidal_potential.cc:
```diff
@@ -0,0 +1,158 @@
+                                  "(which is radial, pointing inward if the gravity "
+                                  "is positive), "
+                                  "and a term that results from a tidal potential and that "
+                                  "leads to a gravity field that varies with latitude and longitude. "
```
```suggestion
                                  "leads to a gravity field that varies with latitude and longitude, as well as time. "
```

bangerth reviewed (commented):
Thank you, and my apologies for taking so long to get to this.

I have only minor comments that you should be able to address in ten minutes or less. Please ping me again when you have made these changes!

gassmoeller reviewed (commented):
Thanks, this looks all good now. Please squash your branch into one commit, push again, and when the testers are done this is ready to be merged. Ping one of us if you need help with the squashing of commits.

gassmoeller commented:
Oh, I overlapped with Wolfgang's review. Please address his comments.

bangerth commented:
For the final version, would you please also squash everything into one commit? Or, if you prefer, one commit for the code changes, one commit for the test case, and commit for the changelog entry. Please talk to any of us how to do that if you don't already know!

hyunseong96 commented:
Hi @bangerth @gassmoeller ,
I finished corrections and squashing commits. I left some replies for comments about \si{year} and T0 and Tstar.
Please let me know if you need things updated!
Thank you :)

bangerth reviewed (approved):
Thank you.

If there is something else you want to do, perhaps copy the formulas about the exact form of the gravity potential and its contribution to gravity from the class documentation to the doc string at the end of radial_with_tidal_potential.cc. But I'm ok with it if you want to do that as a follow-up.

bangerth commented:
Or perhaps even all 6 equations at the top of this page, and a discussion of why you only take the first 2 terms.

bangerth commented:
I think you have some lines with trailing spaces:
```
--- a/source/gravity_model/radial_with_tidal_potential.cc
+++ b/source/gravity_model/radial_with_tidal_potential.cc
@@ -125,9 +125,9 @@ namespace aspect
           const double time_scale = this->get_parameters().convert_to_years ? constants::year_in_seconds : 1.0;
           P = prm.get_double ("Period of nonsynchronous rotation") * time_scale;
 
-          AssertThrow (this->get_parameters().maximum_time_step < P , 
-                                  ExcMessage ("Time step size should be smaller than the period of nonsynchronous rotation. "
-                                                          "Otherwise, the tidal potential will not be resolved properly."));
+          AssertThrow (this->get_parameters().maximum_time_step < P ,
+                       ExcMessage ("Time step size should be smaller than the period of nonsynchronous rotation. "
+                                   "Otherwise, the tidal potential will not be resolved properly."));
         }
         prm.leave_subsection ();
       }
```

hyunseong96 commented:
@bangerth About $T_1$ to $T_3$, they represent the diurnal tidal potential associated with orbital eccentricity and satellite obliquity, and are responsible for the diurnal modulation of the satellite distortion (Tobie et al., 2025). In a nutshell, they are in different time scales in days, as $T_*$ is time-independent and $T_0$ has time scale between 1 kyr and 100kyr. $T_1$ - $T_3$ are usually expressed as tidal heating. Antoniette and I will have more discussion about tidal heating to be expressed more realistically soon using those.

I pushed the correction about indents.  Then, I got the error about using /si{}. I changed into //si{}. please let me know if I changed them wrong.

bangerth commented:
Oh yes, you needed two backslashes. Your change looks correct to me.

As for the other terms: *I* understood what you are saying based on what you say in the comment at the top of this page. But if you look at the manual, you don't have that in front of you: The manual is self-contained, and it would be nice if it contained the explanation you just gave to me.

bangerth commented:
You'll still need this change:
```
@@ -136,7 +136,7 @@ If the function you are describing represents a vector-valued function with mult
 
 **Pattern:** [Double -MAX_DOUBLE...MAX_DOUBLE (inclusive)]
 
-**Documentation:** Mass of body that perturbs gravity of modeled body. The default value is chosen for modeling Europa, therefore, it is the mass of Jupiter. Units is $kg$.
+**Documentation:** Mass of body that perturbs gravity of modeled body. The default value is chosen for modeling Europa, therefore, it is the mass of Jupiter. Units: \si{\kilo\gram}.
 ::::
 
 ::::{dropdown} __Parameter:__ {ref}`Period of nonsynchronous rotation<parameters:Gravity_20model/Radial_20with_20tidal_20potential/Period_20of_20nonsynchronous_20rotation>`
@@ -145,7 +145,7 @@ If the function you are describing represents a vector-valued function with mult
 
 **Pattern:** [Double -MAX_DOUBLE...MAX_DOUBLE (inclusive)]
 
-**Documentation:** Period of nonsynchronous rotation (NSR). The default value is the period of NSR on Europa&rsquo;s icy shell. Units is $year$ when &rsquo;Use years instead of seconds&rsquo; is true, and $second$ when &rsquo;Use years instead of seconds&rsquo; is false.
+**Documentation:** Period of nonsynchronous rotation (NSR). The default value is the period of NSR on Europa&rsquo;s icy shell. Units is year when &rsquo;Use years instead of seconds&rsquo; is true, and \si{\second} when &rsquo;Use years instead of seconds&rsquo; is false.
 ::::
 
 ::::{dropdown} __Parameter:__ {ref}`Semimajor axis of orbit<parameters:Gravity_20model/Radial_20with_20tidal_20potential/Semimajor_20axis_20of_20orbit>`
@@ -154,7 +154,7 @@ If the function you are describing represents a vector-valued function with mult
 
 **Pattern:** [Double -MAX_DOUBLE...MAX_DOUBLE (inclusive)]
 
-**Documentation:** The length of the semimajor axis of the orbit that cause the tidal perturbation. For example, tidal perturbation on Europa happens by Europa orbiting Jupiter, and that on Earth, if Moon is in consideration, happens by Moon orbiting Earth. The default value is for the semimajor axis of Europa&rsquo;s orbit. Units is $m$.
+**Documentation:** The length of the semimajor axis of the orbit that cause the tidal perturbation. For example, tidal perturbation on Europa happens by Europa orbiting Jupiter, and that on Earth, if Moon is in consideration, happens by Moon orbiting Earth. The default value is for the semimajor axis of Europa&rsquo;s orbit. Units: \si{\meter}.
```
I'm sorry it takes so many steps...

hyunseong96 commented:
@bangerth ah thank you for letting me know! I added more explanations on class documentation.
I made the hopeful last md correction. so it will work to the end.

bangerth commented:
Alright, yay!
